integrations/wasi/src/lib.rs

```diff
@@ -0,0 +1,88 @@
+//! A first-party support of the `wasm32-wasip1` target for the **Server-Side**
+//! of Leptos using the [`wasi:http`][wasi-http] proposal.
+//!
+//! [wasi-http]: https://github.com/WebAssembly/wasi-http
+//!
+//! # `Handler`
+//!
+//! The [`prelude::Handler`] is the main abstraction you will use.
+//!
+//! It expects being run in the context of a Future Executor `Task`,
+//! since WASI is, at the moment, a single-threaded environment,
+//! we provide a simple abstraction in the form of [`leptos::spawn::Executor`]
+//! that you can leverage to use this crate.
+//!
+//! ```
+//! use leptos_wasi::{bindings::exports::wasi::http::incoming_handler::Guest, prelude::{IncomingRequest, ResponseOutparam}};
+//!
+//! struct LeptosServer;
+//!
+//! // NB(raskyld): for now, the types to use for the HTTP handlers are the one from
+//! // the `leptos_wasi` crate, not the one generated in your crate.
+//! impl Guest for LeptosServer {
+//!     fn handle(request: IncomingRequest, response_out: ResponseOutparam) {
+//!         // Initiate a single-threaded [`Future`] Executor so we can run the
+//!         // rendering system and take advantage of bodies streaming.
+//!         Executor::init_futures_local_executor().expect("cannot init future executor");
+//!         Executor::spawn(async {
+//!             // declare an async function called `handle_request` and
+//!             // use the Handler in this function.
+//!             handle_request(request, response_out).await;
+//!         });
+//!         Executor::run();
+//!     }
+//! }
+//! ```
+//!
+//! # WASI Bindings
+//!
+//! You are free to use any WIT imports and export any WIT exports but at the moment,
+//! when interacting with this crate, you must use the types that you can find in
+//! this crate [`bindings`].
+//!
+//! You then need to export your implementation using:
+//!
+//! ```
+//! export!(LeptosServer with_types_in leptos_wasi::bindings);
+//! ```
+//!
+//! If you want to use your own bindings for `wasi:http`,
+//! then you need to implement `From` traits
+//! to convert your own bindings into the one in [`bindings`].
+//! Please, note that it will likely implies doing `unsafe`
+//! operations to wrap the resource's `handle() -> u64` in
+//! another type.
+
+#![warn(clippy::all)]
+#![warn(clippy::nursery)]
+#![allow(clippy::restriction)]
+#![deny(clippy::allow_attributes)]
+
+#[allow(warnings)]
+pub mod bindings {
+    wit_bindgen::generate!({
+        path: "wit",
+        pub_export_macro: true,
+        world: "http",
+        generate_all,
+    });
+}
```

Comment:
One thing to consider which may simplify the changes in this PR is directly bringing in the [wasi](https://crates.io/crates/wasi/0.13.1+wasi-0.2.0) crate as a dependency (pinned to `=0.13.1+wasi-0.2.0` may be a good idea) instead of using wit_bindgen directly. The types from the `wasi` crate also have additional trait implementations, like using `std::io::Read/Write` on the stream types 😃 Here's a commit where I updated a draft to use wasi from wit-bindgen https://github.com/brooksmtownsend/reqwest/commit/1a98138b322a944b5ffb496272b0ba6cae783ef2#diff-e3a9842c622287180da65275fb59a80bf4f665098591996b5796f62ff811c41a

I'm not sure if you gave this a shot before, so let me know if this caused issues.

Reply:
That's really cool! And it also solves the issue of what types the user need to use etc to interop with the crate! Will definitely do that :D Thanks Brooks! 💪